src/libmongoc/tests/unified/entity-map.c

```diff
@@ -184,28 +206,61 @@ should_ignore_event (entity_t *client_entity, event_t *event)
    return is_sensitive_command (event);
 }
 
+
+typedef void *(*get_context_func_t) (const void *);
+typedef const bson_t *(*get_command_func_t) (const void *);
+typedef const bson_t *(*get_reply_func_t) (const void *);
+typedef const char *(*get_command_name_func_t) (const void *);
+typedef const char *(*get_database_name_func_t) (const void *);
+typedef const bson_oid_t *(*get_service_id_func_t) (const void *);
+typedef int64_t (*get_server_connection_id_func_t) (const void *);
+
+
+typedef struct command_callback_funcs_t {
+   get_context_func_t get_context;
+   get_command_func_t get_command;
+   get_reply_func_t get_reply;
+   get_command_name_func_t get_command_name;
+   get_database_name_func_t get_database_name;
+   get_service_id_func_t get_service_id;
+   get_server_connection_id_func_t get_server_connection_id;
+} command_callback_funcs_t;
+
+
 static void
-command_started (const mongoc_apm_command_started_t *started)
+observe_event (entity_t *entity,
+               command_callback_funcs_t funcs,
+               const char *type,
+               const void *apm_command)
 {
-   entity_t *entity = NULL;
-   event_t *event = NULL;
+   BSON_ASSERT_PARAM (type);
+   BSON_ASSERT_PARAM (apm_command);
+
+   BSON_ASSERT (funcs.get_context);
+   event_t *const event = event_new (type);
 
-   entity = (entity_t *) mongoc_apm_command_started_get_context (started);
-   event = event_new ("commandStartedEvent");
-   event->command =
-      bson_copy (mongoc_apm_command_started_get_command (started));
-   event->command_name =
-      bson_strdup (mongoc_apm_command_started_get_command_name (started));
+   if (funcs.get_command) {
+      event->command = bson_copy (funcs.get_command (apm_command));
+   }
+
+   if (funcs.get_reply) {
+      event->reply = bson_copy (funcs.get_reply (apm_command));
+   }
 
-   event->database_name =
-      bson_strdup (mongoc_apm_command_started_get_database_name (started));
+   BSON_ASSERT (funcs.get_command_name);
+   event->command_name = bson_strdup (funcs.get_command_name (apm_command));
 
-   if (mongoc_apm_command_started_get_service_id (started)) {
-      bson_oid_copy (mongoc_apm_command_started_get_service_id (started),
-                     &event->service_id);
+   if (funcs.get_database_name) {
+      event->database_name =
+         bson_strdup (funcs.get_database_name (apm_command));
    }
 
-   event->server_connection_id = started->server_connection_id;
+   if (funcs.get_service_id (apm_command)) {
+      bson_oid_copy (funcs.get_service_id (apm_command), &event->service_id);
+   }
+
+   BSON_ASSERT (funcs.get_server_connection_id);
+   event->server_connection_id = funcs.get_server_connection_id (apm_command);
```

Comment:
I am not yet sure why. When `mongoc_apm_command_started_get_server_connection_id` returns `MONGOC_NO_SERVER_CONNECTION_ID` (-1) it is stored as 4294967295. 4294967295 is 2**32 - 1, so I suspected something is casting to a `uint32_t`.

This appears to fail [tests with servers <= 4.0](https://spruce.mongodb.com/task/mongo_c_driver_sanitizers_matrix_asan_asan_sasl_cyrus_openssl_ubuntu1804_clang_test_4.0_sharded_auth_patch_4b1c7751a1b00c3b79f044fe023075ec4a833903_6410a7323066156d84d2c918_23_03_14_16_56_20/logs?execution=0) expecting no server connection ID to be set: `expected no server connectionId, but got 4294967295`

Reply:
Identified cause of failure to be undefined behavior due to incorrect `int64_t` return type for `get_server_connection_id_func_t` where assigned function pointers actually return `int32_t` instead:

```diff
- typedef int64_t (*get_server_connection_id_func_t) (const void *);
+ typedef int32_t (*get_server_connection_id_func_t) (const void *);
```